avogadro/core/molecule.cpp

```diff
@@ -827,6 +827,50 @@ void Molecule::perceiveBondsFromResidueData()
   }
 }
 
+void Molecule::perceiveSubstitutedCations()
```

Comment:
Would this be better suited in `pdbformat` itself? Many formats (SDF, for example) don't need this.

If you wanted a more general form, say `perceiveFormalChargeFromValence()` it might make sense here .. check for both cations and anions .. something like `mdlvalence_p.h` but inverted - formal charges from the bond order.